stdlib/public/core/RangeSet.swift

```diff
@@ -0,0 +1,581 @@
+    let overlapsUpperBound =
+      range.upperBound < _ranges[indices.upperBound - 1].upperBound
+    
+    switch (overlapsLowerBound, overlapsUpperBound) {
+    case (false, false):
+      _ranges.removeSubrange(indices)
+    case (false, true):
+      let newRange =
+        range.upperBound..<_ranges[indices.upperBound - 1].upperBound
+      _ranges.replaceSubrange(indices, with: CollectionOfOne(newRange))
+    case (true, false):
+      let newRange = _ranges[indices.lowerBound].lowerBound..<range.lowerBound
+      _ranges.replaceSubrange(indices, with: CollectionOfOne(newRange))
+    case (true, true):
+      _ranges.replaceSubrange(indices, with: _Pair(
+        _ranges[indices.lowerBound].lowerBound..<range.lowerBound,
+        range.upperBound..<_ranges[indices.upperBound - 1].upperBound
+      ))
+    }
+  }
+}
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+extension RangeSet: Equatable {}
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+extension RangeSet: Hashable where Bound: Hashable {}
+
+// MARK: - Range Collection
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+extension RangeSet {
+  /// A collection of the ranges that make up a range set.
+  public struct Ranges: RandomAccessCollection {
+    var _ranges: _RangeSetStorage<Bound>
+    
+    public var startIndex: Int { _ranges.startIndex }
+    public var endIndex: Int { _ranges.endIndex }
+    
+    public subscript(i: Int) -> Range<Bound> {
+      _ranges[i]
+    }
+  }
+  
+  /// A collection of the ranges that make up the range set.
+  ///
+  /// The ranges that you access by using `ranges` never overlap, are never
+  /// empty, and are always in increasing order.
+  public var ranges: Ranges {
+    Ranges(_ranges: _ranges)
+  }
+}
+
+// MARK: - Collection APIs
+
+@available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *)
+extension RangeSet {
+  /// Creates a new range set containing ranges that contain only the
+  /// specified indices in the given collection.
+  ///
+  /// - Parameters:
+  ///   - index: The index to include in the range set. `index` must be a
+  ///     valid index of `collection` that isn't the collection's `endIndex`.
+  ///   - collection: The collection that contains `index`.
+  public init<S, C>(_ indices: S, within collection: C)
+    where S: Sequence, C: Collection, S.Element == C.Index, C.Index == Bound
+  {
+    for i in indices {
+      self.insert(i, within: collection)
+    }
+  }
+
+  /// Inserts a range that contains only the specified index into the range
+  /// set.
+  ///
+  /// - Parameters:
+  ///   - index: The index to insert into the range set. `index` must be a
+  ///     valid index of `collection` that isn't the collection's `endIndex`.
+  ///   - collection: The collection that contains `index`.
+  ///
+  /// - Complexity: O(*n*), where *n* is the number of ranges in the range
+  ///   set.
+  public mutating func insert<C>(_ index: Bound, within collection: C)
```

Comment:
This is a candidate for effective `inlinable`, depending on performance needs. Doing so will drop generality from Collection to just our Bound. Likely big benefit, write a benchmark to demonstrate.